FxA/FxA/ECDSAKeyPair.m

```diff
@@ -0,0 +1,458 @@
+    }
+    return self;
+}
+
+- (instancetype) initWithBinaryRepresentation: (NSData*) data group: (ECDSAGroup) group;
+{
+    if (group != ECDSAGroupP256) {
+        return self;
+    }
+
+    if ((self = [super init]) != nil) {
+        _ecdsa = EC_KEY_new_by_curve_name(NID_P256);
+
+        EC_GROUP *ecgroup = EC_KEY_get0_group(_ecdsa);
+
+        EC_POINT *pub = EC_POINT_new(ecgroup);
+        if (EC_POINT_oct2point(ecgroup, pub, [data bytes], [data length], NULL)) {
+            EC_KEY_set_public_key(_ecdsa, pub);
+        }
+        EC_POINT_free(pub);
+    }
+    return self;
+}
+
+- (void) dealloc
+{
+    if (_ecdsa != NULL) {
+        EC_KEY_free(_ecdsa);
+        _ecdsa = NULL;
+    }
+}
+
+- (NSDictionary*) JSONRepresentation
+{
+    EC_GROUP *ecgroup = EC_KEY_get0_group(_ecdsa);
+    EC_POINT *pub = EC_KEY_get0_public_key(_ecdsa);
+
+    BIGNUM *x = BN_new();
+    BIGNUM *y = BN_new();
+    if (!EC_POINT_get_affine_coordinates_GFp(ecgroup, pub, x, y, NULL)) {
+        BN_free(x);
+        BN_free(y);
+        return nil;
+    }
+
+    return @{
+             @"algorithm": @"ES",
+             @"x": [[CHNumber numberWithOpenSSLNumber: x] stringValue],
+             @"y": [[CHNumber numberWithOpenSSLNumber: y] stringValue]
+             };
+}
+
+// This representation is the standard binary representation of an elliptic curve point
+//
+//    point = 0x04 || x[32] || y[32]
+//
+// Where the coordinate size (32) is dictated by the use of P-256.  If different curves are
+// supported in the future, the will produce differently-sized public keys.
+- (NSData*) BinaryRepresentation
+{
+    EC_GROUP *ecgroup = EC_KEY_get0_group(_ecdsa);
+    EC_POINT *pub = EC_KEY_get0_public_key(_ecdsa);
+
+    size_t ptLen = EC_POINT_point2oct(ecgroup, pub, POINT_CONVERSION_UNCOMPRESSED, NULL, 0, NULL);
+
+    NSData *mut = [[NSMutableData alloc] initWithLength: ptLen];
+    EC_POINT_point2oct(ecgroup, pub, POINT_CONVERSION_UNCOMPRESSED, [mut bytes], ptLen, NULL);
+    return mut;
+}
+
+- (NSString*) algorithm
+{
+    return [[NSString alloc] initWithUTF8String:kECDSAP256Algorithm];
+}
+
+- (BOOL) verifySignature: (NSData*) signature againstMessage: (NSData*) message
+{
+    BOOL verified = NO;
+    
+    EVP_PKEY *pkey = EVP_PKEY_new();
+    if (pkey != NULL) {
+        if (EVP_PKEY_set1_EC_KEY(pkey, _ecdsa) == 1) {
+            EVP_MD_CTX *ctx = EVP_MD_CTX_create();
+            if (ctx != NULL) {
+                if (EVP_VerifyInit_ex(ctx, EVP_sha256(), NULL) == 1) {
+                    if (EVP_VerifyUpdate(ctx, [message bytes], [message length]) ==  1) {
+                        int err = EVP_VerifyFinal(ctx, [signature bytes], [signature length], pkey);
+                        if (err == 1) {
+                            verified = YES;
+                        } else if (err == -1) {
+                            unsigned long e = ERR_get_error();
+                            
+                            char buf[120];
```

Comment:
[OpenSSL Documentation](https://www.openssl.org/docs/manmaster/crypto/ERR_error_string.html) defines this `buf` as being at least 256 bytes long.

If you want to use the shorter buffer, please use `ERR_error_string_n`